research/gaia/gremlin/gs_gremlin/src/graph_proxy/global_storage.rs

```diff
@@ -0,0 +1,462 @@
+        let label_ids = encode_storage_label(params.labels.as_ref(), schema.clone());
+        let prop_ids = encode_storage_prop_key(params.props.as_ref(), schema.clone());
+        let filter = params.filter.clone();
+        let result = store
+            .get_all_edges(
+                si,
+                label_ids.as_ref(),
+                None,
+                None,
+                prop_ids.as_ref(),
+                params.limit.unwrap_or(0),
+                // TODO(bingqing): optimization, each worker may scan some partitions on current server.
+                &vec![],
+            )
+            .map(move |e| to_runtime_edge(&e));
+
+        Ok(filter_limit!(result, filter, None))
+    }
+
+    fn get_vertex(
+        &self,
+        ids: &[ID],
+        params: &QueryParams<Vertex>,
+    ) -> DynResult<Box<dyn Iterator<Item = Vertex> + Send>> {
+        let store = self.store.clone();
+        let si = params
+            .get_extra_param(SNAPSHOT_ID)
+            .ok_or(str_to_dyn_error("get snapshot_id failed"))?
+            .as_i64()
+            .map_err(|e| str_to_dyn_error(&e.to_string()))? as SnapshotId;
+        let schema = store
+            .get_schema(si)
+            .ok_or(str_to_dyn_error("get schema failed"))?;
+        let prop_ids = encode_storage_prop_key(params.props.as_ref(), schema.clone());
+        let filter = params.filter.clone();
+        let partition_label_vertex_ids =
+            build_partition_label_vertex_ids(ids, self.partition_manager.clone());
+        let result = store
+            .get_vertex_properties(si, partition_label_vertex_ids, prop_ids.as_ref())
+            .map(move |v| to_runtime_vertex(&v));
+
+        Ok(filter_limit!(result, filter, None))
+    }
+
+    fn get_edge(
+        &self,
+        _ids: &[ID],
+        _params: &QueryParams<Edge>,
+    ) -> DynResult<Box<dyn Iterator<Item = Edge> + Send>> {
+        // TODO(bingqing): adapt get_edge when graphscope support this
+        Err(str_to_dyn_error(
+            "GraphScope storage does not support get_edge for now",
+        ))
+    }
+
+    fn prepare_explore_vertex(
+        &self,
+        direction: Direction,
+        params: &QueryParams<Vertex>,
+    ) -> DynResult<Box<dyn Statement<ID, Vertex>>> {
+        let filter = params.filter.clone();
+        let limit = params.limit.clone();
+        let store = self.store.clone();
+        let partition_manager = self.partition_manager.clone();
+        let si = params
+            .get_extra_param(SNAPSHOT_ID)
+            .ok_or(str_to_dyn_error("get snapshot_id failed"))?
+            .as_i64()
+            .map_err(|e| str_to_dyn_error(&e.to_string()))? as SnapshotId;
+        let schema = store
+            .get_schema(si)
+            .ok_or(str_to_dyn_error("get schema failed"))?;
+        let edge_label_ids = encode_storage_label(params.labels.as_ref(), schema.clone());
+
+        let stmt = from_fn(move |v: ID| {
```

Comment:
My concern: calling build_partition_vertex_ids for each vertex may not be a good idea, where build_partition_vertex_ids is for a bulk of vertices and will have to new a hashmap each time. This is too heavy for processing a single vertex. Can have better design! This applies to the following codes.

Reply:
Totally agree. However, GlobalGraphQuery trait defined in GraphScope is designed to process a bulk of vertices, while in GAIA, the bulk process is actually in engine, and adapter is only able to process a single vertex each time.

Reply:
Totally agree. However, GlobalGraphQuery trait defined in GraphScope is designed to process a bulk of vertices, while in GAIA, the bulk process is actually in engine, and adapter is only able to process a single vertex each time.
In the new commit, I remove the unnecessary new HashMap in build_partition_vertex_ids function.